business/clustering.go

```diff
@@ -0,0 +1,160 @@
+package business
+
+import (
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+)
+
+// ClusteringService is a support service for retrieving data about the mesh environment
+// when Istio is installed with multi-cluster enabled.
+type ClusteringService struct {
+	k8s           kubernetes.ClientInterface
+	businessLayer *Layer
+}
+
+// MeshCluster holds some metadata about a cluster that is
+// part of the mesh.
+type MeshCluster struct {
+	// ApiEndpoint is the URL where the Kubernetes/Cluster API Server can be contacted
+	ApiEndpoint string `json:"apiEndpoint"`
+
+	// IsHomeCluster specifies if this cluster is hosting Kiali (and the observed Mesh Control Plane)
+	IsHomeCluster bool `json:"isHomeCluster"`
+
+	// Name specifies the CLUSTER_ID as known by the Control Plane
+	Name string `json:"name"`
+
+	// SecretName is the name of the kubernetes "remote secret" where data of this cluster was resolved
+	SecretName string `json:"secretName"`
+}
+
+// GetMeshClusters resolves the Kubernetes clusters that are hosting the mesh. Resolution
+// is done as best-effort using the resources that are present in the cluster.
+func (in *ClusteringService) GetMeshClusters() ([]MeshCluster, error) {
+	var err error
+
+	remoteClusters, err := in.resolveRemoteClustersFromSecrets()
+	if err != nil {
+		return nil, err
+	}
+
+	myCluster, err := in.resolveMyControlPlaneCluster()
+	if err != nil {
+		return nil, err
+	}
+
+	allClusters := append(remoteClusters, *myCluster)
+
+	return allClusters, nil
+}
+
+// resolveMyControlPlaneCluster tries to resolve the metadata about the cluster where
+// Kiali is installed. This assumes that the mesh Control Plane is installed in the
+// same cluster as Kiali.
+func (in *ClusteringService) resolveMyControlPlaneCluster() (*MeshCluster, error) {
```

Comment:
Again, we could just be explicit, and instead of inferring the Kiali point of view in the API, just resolveKialiControlPlaneCluster().

Reply:
OK, I'll rename this.